### PR TITLE
#3155 Enable json refiner to check input data according to definition

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #3155)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala [hunk 1]
```diff
@@ -159,33 +156,30 @@ class JDBCSourceConnector extends RowSourceConnector {
...
+      .optional(JDBC_FLUSHDATA_SIZE_DEFAULT)
       .orderInGroup(counter.getAndIncrement())
       .build(),
     SettingDef
       .builder()
       .displayName("Fetch data frequence")
       .documentation("Setting fetch data frequency from database")
-      .valueType(SettingDef.Type.DURATION)
       .key(JDBC_FREQUENCE_TIME)
       .optional(String.valueOf(JDBC_FREQUENCE_TIME_DEFAULT))
```
Comment (jackyoh) [inline ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala]: 這一段我很像寫錯了，要改成：
.optional(toJavaDuration(JDBC_FREQUENCE_TIME_DEFAULT))

File: ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala [hunk 2]
```diff
@@ -192,73 +192,103 @@ class TestJDBCSourceConnectorDefinition extends WithBrokerWorker with Matchers {
...
+      .necessary() shouldBe Necessary.REQUIRED
 
     response
       .settings()
       .asScala
       .filter(_.definition().key() == ConnectorDefUtils.NUMBER_OF_TASKS_DEFINITION.key())
       .head
       .definition()
-      .required() shouldBe true
+      .necessary() shouldBe Necessary.REQUIRED
```
Comment (jackyoh) [inline ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala]: 這一段跑測試結果會是 OPTIONAL_WITH_DEFAULT，不太確定要在哪裡修改

File: ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala [hunk 3]
```diff
@@ -93,7 +93,7 @@ class TestJDBCSourceConnectorDefinition extends WithBrokerWorker with Matchers {
   @Test
   def checkFrequenceTime(): Unit = {
     val definition = jdbcSource.definitions().asScala.find(_.key() == JDBC_FREQUENCE_TIME).get
-    definition.required shouldBe false
+    definition.necessary should not be Necessary.REQUIRED
     definition.defaultValue shouldBe String.valueOf(JDBC_FREQUENCE_TIME_DEFAULT)
```
Comment (jackyoh) [inline ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala]: 這邊的預期結果很像會改成 PT0S

File: ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala (hunk 1 above)
Comment (chia7712) [reply ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala]: the string of scala duration is ok to ohara :)

File: ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala (hunk 2 above)
Comment (chia7712) [reply ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala]: nice finding. I will fix it in next commit

File: ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala (hunk 3 above)
Comment (chia7712) [reply ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala]: ditto :)

File: ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala (hunk 2 above)
Comment (jackyoh) [reply ohara-connector/src/test/scala/com/island/ohara/connector/jdbc/source/TestJDBCSourceConnectorDefinition.scala]: Thank you.

File: ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala (hunk 1 above)
Comment (jackyoh) [reply ohara-connector/src/main/scala/com/island/ohara/connector/jdbc/source/JDBCSourceConnector.scala]: 了解

**Comment by chia7712** [comment]
retry k8sConfigurator

**Comment by chia7712** [comment]
@wu87988622 @konekoya @saivirtue Please take a look at the APIs failure :(

**Comment by chia7712** [comment]
retry k8sConfigurator

**Comment by jackyoh** [comment]
retry manager-api

**Comment by konekoya** [comment]
@chia7712 @jackyoh should we change the failing notification from Jenkins from pinging specific fontend devs like (@wu87988622 and me) to pinging @oharastream/frontend? I think the latter is more flexible?

**Comment by jackyoh** [comment]
> @chia7712 @jackyoh should we change the failing API notification from Jenkins from pinging specific fontend devs like (@wu87988622 and me) to pinging @oharastream/frontend? I think the latter is more flexible?

在 jenkins 上應該可以修改，我來嘗試修改

**Comment by chia7712** [comment]
@oharastream/frontend any feedback?

**Comment by konekoya** [comment]
> @oharastream/frontend any feedback?

@wu87988622 will look into it :)

File: docs/src/setting_definition.rst [hunk 4]
```diff
@@ -496,3 +554,24 @@ a checker which validate whether the input value is able to be cast to
...
+Currently, blacklist is used by Array type only.
+
+.. _recommendedValues:
+
+Recommended values
+------------------
+
+Recommended values is used by Ohara UI that it able to pop a list to users when they are using UI.
+
+Currently, blacklist is used by String type only.
```
Comment (saivirtue) [inline docs/src/setting_definition.rst]: bkacklist => recommend?

Comment (chia7712) [reply docs/src/setting_definition.rst]: nice finding :)

**Comment by chia7712** [comment]
@wu87988622 Do you want to apply "Necessary" in this PR? I had filed a issue (#3166) for that.

**Comment by chia7712** [comment]
BTW, which one is broken?

**Comment by wu87988622** [comment]
> BTW, which one is broken?

I have updated the new API changes to this pr

**Comment by chia7712** [comment]
> I have updated the new API changes to this pr

which one?

**Comment by wu87988622** [comment]
> > I have updated the new API changes to this pr
> 
> which one?

All APIs using definition

**Comment by chia7712** [comment]
> All APIs using definition

I fail to catch the broken stuff. Or why the APIs tests fails?

**Comment by wu87988622** [comment]
> > All APIs using definition
> 
> I fail to catch the broken stuff. Or why the APIs tests fails?

Because some key words were used in API tests before,
Like **required**
Now it has been replaced with **necessary** in this pr

**Comment by chia7712** [comment]
> Like required
Now it has been replaced with necessary in this pr

"required" is not removed. That is why I opened #3166

**Comment by wu87988622** [comment]
> > Like required
> > Now it has been replaced with necessary in this pr
> 
> "required" is not removed. That is why I opened #3166

oh!! you can close #3166 
manager API is ready.

**Comment by chia7712** [comment]
我想要知道的事情是，就我的測試上應該沒有破壞舊版的APIs，但因為APIs測試失敗，所以我想知道是哪個部份破壞了APIs

**Comment by wu87988622** [comment]
> 我想要知道的事情是，就我的測試上應該沒有破壞舊版的APIs，但因為APIs測試失敗，所以我想知道是哪個部份破壞了APIs

轉換DefaultVale的部分，之前沒有明確的定義該如何轉換，現在已經將轉換邏輯更新了。

**Comment by chia7712** [comment]
> 轉換DefaultVale的部分，之前沒有明確的定義該如何轉換，現在已經將轉換邏輯更新了。

不懂。為何之前不會失敗，這支PR就會失敗？在default value有什麼變化嗎

**Comment by wu87988622** [comment]
> > 轉換DefaultVale的部分，之前沒有明確的定義該如何轉換，現在已經將轉換邏輯更新了。
> 
> 不懂。為何之前不會失敗，這支PR就會失敗？在default value有什麼變化嗎

之前過度版本只有針對數字做轉換，其他都不會轉換。所以只要出現例外就不處理了
在變化上的話不是很確定是哪個值的改變

**Comment by chia7712** [comment]
> 在變化上的話不是很確定是哪個值的改變

沒有log可以看嗎？這是個很重要的點，因為有東西通過了後端一狗票測試且被破壞了

**Comment by wu87988622** [comment]
> > 在變化上的話不是很確定是哪個值的改變
> 
> 沒有log可以看嗎？這是個很重要的點，因為有東西通過了後端一狗票測試且被破壞了

那等我一下，我去找出關鍵點～

**Comment by wu87988622** [comment]
> > 在變化上的話不是很確定是哪個值的改變
> 
> 沒有log可以看嗎？這是個很重要的點，因為有東西通過了後端一狗票測試且被破壞了

經過比對
兩個版本zookeeper中的definition定義得數量不一樣
舊版
```
{"imageName":"oharastream/zookeeper:0.9.0-SNAPSHOT","settingDefinitions":[{"reference":"NONE","displayName":"peerPort","internal":false,"documentation":"the port exposed to each quorum","valueType":"PORT","tableKeys":[],"orderInGroup":5,"key":"peerPort","required":false,"defaultValue":null,"group":"core","editable":true},{"reference":"NONE","displayName":"electionPort","internal":false,"documentation":"quorum leader election port","valueType":"PORT","tableKeys":[],"orderInGroup":6,"key":"electionPort","required":false,"defaultValue":null,"group":"core","editable":true},{"reference":"NONE","displayName":"group","internal":false,"documentation":"group of this worker cluster","valueType":"STRING","tableKeys":[],"orderInGroup":0,"key":"group","required":false,"defaultValue":"default","group":"core","editable":true},{"reference":"NONE","displayName":"name","internal":false,"documentation":"name of this worker cluster","valueType":"STRING","tableKeys":[],"orderInGroup":1,"key":"name","required":false,"defaultValue":null,"group":"core","editable":true},{"reference":"NONE","displayName":"initLimit","internal":false,"documentation":"timeout to connect to leader","valueType":"INT","tableKeys":[],"orderInGroup":8,"key":"initLimit","required":false,"defaultValue":"10","group":"core","editable":true},{"reference":"NONE","displayName":"tickTime","internal":false,"documentation":"basic time unit in zookeeper","valueType":"INT","tableKeys":[],"orderInGroup":7,"key":"tickTime","required":false,"defaultValue":"2000","group":"core","editable":true},{"reference":"NONE","displayName":"clientPort","internal":false,"documentation":"the port exposed to client to connect to zookeeper","valueType":"PORT","tableKeys":[],"orderInGroup":4,"key":"clientPort","required":false,"defaultValue":null,"group":"core","editable":true},{"reference":"NONE","displayName":"nodeNames","internal":false,"documentation":"the nodes hosting this cluster","valueType":"ARRAY","tableKeys":[],"orderInGroup":2,"key":"nodeNames","required":true,"defaultValue":null,"group":"core","editable":true},{"reference":"NONE","displayName":"dataDir","internal":false,"documentation":"the folder used to store zookeeper data","valueType":"STRING","tableKeys":[],"orderInGroup":10,"key":"dataDir","required":false,"defaultValue":"/home/ohara/default/data","group":"core","editable":true},{"reference":"NONE","displayName":"syncLimit","internal":false,"documentation":"the out-of-date of a sever from leader","valueType":"INT","tableKeys":[],"orderInGroup":9,"key":"syncLimit","required":false,"defaultValue":"5","group":"core","editable":true},{"reference":"NONE","displayName":"tags","internal":false,"documentation":"the tags to this cluster","valueType":"TAGS","tableKeys":[],"orderInGroup":3,"key":"tags","required":false,"defaultValue":null,"group":"core","editable":true}]}
```
新版
```
{"imageName":"oharastream/zookeeper:0.9.0-SNAPSHOT","settingDefinitions":[{"blacklist":[],"reference":"NONE","displayName":"peerPort","internal":false,"documentation":"the port exposed to each quorum","necessary":"OPTIONAL_WITH_RANDOM_DEFAULT","valueType":"BINDING_PORT","tableKeys":[],"orderInGroup":7,"key":"peerPort","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"jmxPort","internal":false,"documentation":"the port used to expose the metrics of this cluster","necessary":"OPTIONAL_WITH_RANDOM_DEFAULT","valueType":"BINDING_PORT","tableKeys":[],"orderInGroup":4,"key":"jmxPort","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"group","internal":false,"documentation":"group of this worker cluster","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"STRING","tableKeys":[],"orderInGroup":0,"key":"group","required":false,"defaultValue":"default","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"electionPort","internal":false,"documentation":"quorum leader election port","necessary":"OPTIONAL_WITH_RANDOM_DEFAULT","valueType":"BINDING_PORT","tableKeys":[],"orderInGroup":8,"key":"electionPort","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"name","internal":false,"documentation":"name of this worker cluster","necessary":"OPTIONAL_WITH_RANDOM_DEFAULT","valueType":"STRING","tableKeys":[],"orderInGroup":1,"key":"name","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":["start","stop","pause","resume"],"reference":"NONE","displayName":"nodeNames","internal":false,"documentation":"the nodes hosting this cluster","necessary":"REQUIRED","valueType":"ARRAY","tableKeys":[],"orderInGroup":5,"key":"nodeNames","required":true,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"tickTime","internal":false,"documentation":"basic time unit in zookeeper","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"INT","tableKeys":[],"orderInGroup":9,"key":"tickTime","required":false,"defaultValue":"2000","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"clientPort","internal":false,"documentation":"the port used to expose the service","necessary":"OPTIONAL_WITH_RANDOM_DEFAULT","valueType":"BINDING_PORT","tableKeys":[],"orderInGroup":3,"key":"clientPort","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"initLimit","internal":false,"documentation":"timeout to connect to leader","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"INT","tableKeys":[],"orderInGroup":10,"key":"initLimit","required":false,"defaultValue":"10","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"imageName","internal":false,"documentation":"the docker image of this service","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"STRING","tableKeys":[],"orderInGroup":2,"key":"imageName","required":false,"defaultValue":"oharastream/zookeeper:0.9.0-SNAPSHOT","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"dataDir","internal":false,"documentation":"the folder used to store zookeeper data","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"STRING","tableKeys":[],"orderInGroup":12,"key":"dataDir","required":false,"defaultValue":"/home/ohara/default/data","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"syncLimit","internal":false,"documentation":"the out-of-date of a sever from leader","necessary":"OPTIONAL_WITH_DEFAULT","valueType":"INT","tableKeys":[],"orderInGroup":11,"key":"syncLimit","required":false,"defaultValue":"5","recommendedValues":[],"group":"core","editable":true},{"blacklist":[],"reference":"NONE","displayName":"tags","internal":false,"documentation":"the tags to this cluster","necessary":"OPTIONAL","valueType":"TAGS","tableKeys":[],"orderInGroup":6,"key":"tags","required":false,"defaultValue":null,"recommendedValues":[],"group":"core","editable":true}]}
```
其中關鍵就是imageName
舊版沒有定義該項目 所以後端幫我們填入
新版有定義該項目 所以我們透過defaultVale轉換填入
我們前端這邊轉換上是不齊全的

結論就是....
前端的錯 :(

**Comment by chia7712** [comment]
> 其中關鍵就是imageName
舊版沒有定義該項目 所以後端幫我們填入
新版有定義該項目 所以我們透過defaultVale轉換填入
我們前端這邊轉換上是不齊全的

感謝。

這一點蠻有趣的，因為之後的相容性保證是舊的會保持，新的會加入。

換言之新加入的參數只要不是required，那麼就"不應該"導致APIs fails。例如我們在zk APIs上又加入一組參數來控制xxx，該參數有default value，那麼該參數的加入就1.) 不會破壞APIs測試 2)不會破壞UI

我們必須盡力保持上述的特性，否則definition就跟沒有一樣，因為完全沒有讓前後端脫鉤

**Comment by chia7712** [comment]
retry configurator

**Comment by chia7712** [comment]
retry k8sConfigurator

**Comment by chia7712** [comment]
retry configurator

**Pull request merged**